### PR TITLE
Stream Bedrock responses

### PR DESCRIPTION
**Pull request by sobolk**
## Problem

For some the time to first response matters. I.e. this PR adds support for streaming API in the conversation handler.

## Changes

1. Adds `streamResponse` property in the incoming event. This is used to select desired behavior.
2. Add implementation of streaming bedrock responses. Inspired by https://docs.aws.amazon.com/bedrock/latest/userguide/tool-use-examples.html (see `ConverseStream` section)
3. Modifies unit and e2e tests to run common set of scenarios for both response emission mechanisms.
4. Adds couple of tests for new logical branches.

Note: most of the volume in this PR are test adjustments. Please see comments I left below.

## Streaming vs non-streaming

With this change we'll support both ways of interacting with Bedrock.

Why both?
1. Using streaming vs non-streaming is a tradeoff between time to first chunk vs overall cost of interaction. In other words, with streaming customer can achieve a perception of fluent interaction, but it involves more network transfers and intermediate storage upstream to be able to track the conversation.
2. The variety of content blocks that Bedrock can emit is different streaming vs non-streaming. E.g. streaming only supports text and tools, other content blocks are only available in non-streaming calls.
3. Bedrock seems to be "talking" about tools it's using when streaming. Some may find it unwanted.

## Validation

1. Added new unit tests.
4. Modified existing unit and e2e tests to cover new emission mechanism.
6. Snapshot release has been shipped and integrated and manually tested e2e with upstream components.

Video from e2e testing by @dbanksdesign :

https://github.com/user-attachments/assets/7fed3f25-a6b3-458d-a573-7438ea961656

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.test.ts [hunk 1]
```diff
@@ -70,315 +75,702 @@ void describe('Bedrock converse adapter', () => {
...
   );
 
-  void it('calls bedrock to get conversation response', async () => {
-    const event: ConversationTurnEvent = {
-      ...commonEvent,
-    };
+  [false, true].forEach((streamResponse) => {
+    // This is a common set of use cases that both streaming and non-streaming version must support.
+    void describe(`${streamResponse ? 'with' : 'without'} streaming`, () => {
+      void it('calls bedrock to get conversation response', async () => {
```
Comment (sobolk) [inline packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.test.ts:81]: The majority of changes here are to adapt existing tests to handle both streaming and non streaming calls.

The reason for this is that:
1. we want to have functional parity between streaming and non-streaming
2. at least 2/3 is common - e.g. mocking bedrock inputs.
3. the form of output is different, but it carries more-less same content.

These changes contribute most to the volume of this PR.

File: packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.test.ts [hunk 2]
```diff
@@ -451,396 +843,6 @@ void describe('Bedrock converse adapter', () => {
     );
   });
 
-  void it('executable tool error is reported to bedrock', async () => {
```
Comment (sobolk) [inline packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.test.ts:454]: Some tests are moved up and down to accommodate new grouping. They're NOT deleted.

File: packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.ts [hunk 3]
```diff
@@ -153,6 +157,169 @@ export class BedrockConverseAdapter {
     return bedrockResponse.output?.message?.content ?? [];
   };
 
+  /**
+   * Asks Bedrock for response using streaming version of Converse API.
+   */
+  async *askBedrockStreaming(): AsyncGenerator<StreamingResponseChunk> {
+    const { modelId, systemPrompt, inferenceConfiguration } =
+      this.event.modelConfiguration;
```
Comment (sobolk) [inline packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.ts:165]: This is the core of new functionality.

Inspired by https://docs.aws.amazon.com/bedrock/latest/userguide/tool-use-examples.html (make sure to look at `ConverseStream` section). 

It is using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function* and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator .
to represent stream of asynchronous chunks.

File: packages/ai-constructs/src/conversation/runtime/conversation_turn_executor.ts [hunk 4]
```diff
@@ -31,9 +35,16 @@ export class ConversationTurnExecutor {
...
       this.logger.debug('Event received:', this.event);
 
-      const assistantResponse = await this.bedrockConverseAdapter.askBedrock();
-
-      await this.responseSender.sendResponse(assistantResponse);
+      if (this.event.streamResponse) {
+        const chunks = this.bedrockConverseAdapter.askBedrockStreaming();
+        for await (const chunk of chunks) {
+          await this.streamingResponseSender.sendResponseChunk(chunk);
+        }
```
Comment (sobolk) [inline packages/ai-constructs/src/conversation/runtime/conversation_turn_executor.ts:42]: This emits chunks as they come asynchronously.

File: packages/integration-tests/src/test-projects/conversation-handler/amplify/custom-conversation-handler/custom_handler.ts [hunk 5]
```diff
@@ -21,7 +21,7 @@ const thermometer = createExecutableTool(
   },
   (input) => {
     const city = input.city;
-    if (city === 'Seattle' || city === 'Boston' || city === 'Miami') {
+    if (city === 'Seattle' || city === 'Boston') {
```
Comment (sobolk) [inline packages/integration-tests/src/test-projects/conversation-handler/amplify/custom-conversation-handler/custom_handler.ts:24]: This is unrelated to rest of PR.

I removed one of the cities because:
1. two is enough to prove that multiple tools work in single turn
2. having 3 tools makes Bedrock hallucinate empty answer more frequently causing test instabilities.

File: packages/integration-tests/src/test-project-setup/conversation_handler_project.ts [hunk 6]
```diff
@@ -223,14 +209,24 @@ class ConversationHandlerTestProject extends TestProjectBase {
       backendId,
       authenticatedUserCredentials.accessToken,
       clientConfig.data.url,
-      apolloClient
+      apolloClient,
+      false
```
Comment (sobolk) [inline packages/integration-tests/src/test-project-setup/conversation_handler_project.ts:213]: Changes is in this file are to exercise same scenarios with and without streaming.

**Review by Amplifiyer — APPROVED** [review]

**Review by rtpascual — APPROVED** [review]
